exchanges/websocket/wshandler/wshandler.go

```diff
@@ -204,124 +178,121 @@ func (w *Websocket) Shutdown() error {
 		w.Orderbook.FlushCache()
 		w.m.Unlock()
 	}()
-	if !w.connected && w.ShutdownC == nil {
+	if !w.IsConnected() {
 		return fmt.Errorf("%v cannot shutdown a disconnected websocket", w.exchangeName)
 	}
 	if w.verbose {
 		log.Debugf(log.WebsocketMgr, "%v shutting down websocket channels", w.exchangeName)
 	}
-	timer := time.NewTimer(15 * time.Second)
-	c := make(chan struct{}, 1)
-
-	go func(c chan struct{}) {
-		close(w.ShutdownC)
-		w.Wg.Wait()
-		if w.verbose {
-			log.Debugf(log.WebsocketMgr, "%v completed websocket channel shutdown", w.exchangeName)
-		}
-		c <- struct{}{}
-	}(c)
-
-	select {
-	case <-c:
-		w.connected = false
-		return nil
-	case <-timer.C:
-		return fmt.Errorf("%s websocket routines failed to shutdown after 15 seconds",
-			w.GetName())
-	}
-}
-
-// WebsocketReset sends the shutdown command, waits for channel/func closure and then reconnects
-func (w *Websocket) WebsocketReset() {
-	err := w.Shutdown()
-	if err != nil {
-		// does not return here to allow connection to be made if already shut down
-		w.DataHandler <- fmt.Errorf("%v shutdown error: %v", w.exchangeName, err)
-	}
-	log.Infof(log.WebsocketMgr, "%v reconnecting to websocket", w.exchangeName)
-	w.m.Lock()
-	w.init = true
-	w.m.Unlock()
-	err = w.Connect()
-	if err != nil {
-		w.DataHandler <- fmt.Errorf("%v connection error: %v", w.exchangeName, err)
+	close(w.ShutdownC)
+	w.Wg.Wait()
+	w.setConnectedStatus(false)
+	w.setConnectingStatus(false)
+	if w.verbose {
+		log.Debugf(log.WebsocketMgr, "%v completed websocket channel shutdown", w.exchangeName)
 	}
+	return nil
 }
 
-// trafficMonitor monitors traffic and switches connection modes for websocket
+// trafficMonitor uses a timer of WebsocketTrafficLimitTime and once it expires
+// Will reconnect if the TrafficAlert channel has not received any data
+// The trafficTimer will reset on each traffic alert
 func (w *Websocket) trafficMonitor(wg *sync.WaitGroup) {
 	w.Wg.Add(1)
-	wg.Done() // Makes sure we are unlocking after we add to waitgroup
+	wg.Done()
+	trafficTimer := time.NewTimer(w.trafficTimeout)
+
 	defer func() {
-		if w.connected {
-			w.Disconnected <- struct{}{}
-		}
+		trafficTimer.Stop()
 		w.Wg.Done()
 	}()
 
-	// Define an initial traffic timer which will be a delay then fall over to
-	// WebsocketTrafficLimitTime after first response
-	trafficTimer := time.NewTimer(5 * time.Second)
 	for {
 		select {
-		case <-w.ShutdownC: // Returns on shutdown channel close
+		case <-w.ShutdownC:
 			if w.verbose {
 				log.Debugf(log.WebsocketMgr, "%v trafficMonitor shutdown message received", w.exchangeName)
 			}
 			return
-		case <-w.TrafficAlert: // Resets timer on traffic
-			w.m.Lock()
-			if !w.connected {
-				w.Connected <- struct{}{}
-				w.connected = true
+		case <-w.TrafficAlert:
+			if !trafficTimer.Stop() {
+				<-trafficTimer.C
```

Comment:
ch ->
select {
case <-trafficTimer.C:
default:
}

also defer like this as well.